### PR TITLE
(PUP-7433) Reorganize Gemfile

### PR DESCRIPTION
**Pull request by joshcooper**
Previously we excluded "development extra packaging" during CI runs,
which included some unnecessary gems from the "test" group, such as
ronn, yard, memory_profiler. And it excluded msgpack, which is an
optional feature-based dependency, which means tests were being skipped.

This PR reorganizes groups into:

    features
    test
    development*
    packaging
    documentation

where (*) means the group is optional and will only be installed when
requested as `bundle install --with <name>`.

For local testing, run `bundle install`.
For development, `bundle install --with development`.
To generate documentation like man pages, `bundle install`.
Travis/Appveyor exclude the `packaging` and `documentation` groups since they are not needed there.

This PR preserves all of the gems and constraints, except for `puppetlabs_spec_helper`, which was unused and is now removed.

### CODE REVIEW

**Comment by joshcooper** [comment]
Need to make a pass through `Puppet.features.<name>?` and see if any additional gems should be added to the `features` group like `net-telnet`. /cc @branan

File: Gemfile [hunk 1]
```diff
@@ -4,77 +4,60 @@ gemspec
...
+    gem 'ruby-prof', '>= 0.16.0', require: false
   end
-
-  gem 'gettext-setup', '~> 0.28', :require => false
 end
 
-group(:extra) do
-  gem "puppetlabs_spec_helper", :require => false
-  gem "msgpack", :require => false
+group(:packaging, optional: true) do
```
Comment (joshcooper) [inline Gemfile]: probably need to leave packaging as non-optional so we don't need to change CI pipelines.

File: Rakefile [hunk 2]
```diff
@@ -23,11 +23,13 @@ require 'open3'
 
 Dir['tasks/**/*.rake'].each { |t| load t }
 
-begin
-  require 'packaging'
-  Pkg::Util::RakeUtils.load_packaging_tasks
-rescue LoadError => e
-  puts "Error loading packaging rake tasks: #{e}"
+if Rake.application.top_level_tasks.grep(/^(pl:|package:)/).any?
```
Comment (Magisus) [inline Rakefile]: +1 to this, just ran into an issue where bundler didn't install the packaging gems by default, but also couldn't load any rake tasks at all without it.

**Comment by joshcooper** [comment]
jenkins please test this

**Comment by joshcooper** [comment]
jenkins please test this on windows2012r2-64a

File: Gemfile [hunk 3]
```diff
@@ -4,77 +4,67 @@ gemspec
...
-  gem "puppetlabs_spec_helper", :require => false
-  gem "msgpack", :require => false
+group(:packaging) do
+  gem 'packaging', *location_for(ENV['PACKAGING_LOCATION'] || '~> 0.99')
 end
 
+group(:documentation, optional: true) do
+  gem 'gettext-setup', '~> 0.28', require: false, platforms: [:ruby]
+  gem 'ronn', '~> 0.7.3', require: false, platforms: [:ruby]
+end
```
Comment (joshcooper) [inline Gemfile]: need to update ci-job-configs to include the `documentation` group

**Comment by joshcooper** [comment]
jenkins please test this

**Pull request merged**